spatialos-sdk/src/worker/component.rs

```diff
@@ -1,35 +1,463 @@
+                    component_id: request.component_id,
+                    internal: request.schema_type,
+                },
+                user_handle: request.user_handle,
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct CommandResponse {
+        pub component_id: ComponentId,
+        pub schema_type: SchemaCommandResponse,
+        pub user_handle: *mut Worker_CommandResponseHandle,
+    }
+
+    impl From<&Worker_CommandResponse> for CommandResponse {
+        fn from(response: &Worker_CommandResponse) -> Self {
+            CommandResponse {
+                component_id: response.component_id,
+                schema_type: SchemaCommandResponse {
+                    component_id: response.component_id,
+                    internal: response.schema_type,
+                },
+                user_handle: response.user_handle,
+            }
         }
     }
 }
 
-// TODO: Wrap Schema_ComponentUpdate
-pub struct ComponentUpdate {
-    pub component_id: ComponentId,
-    pub schema_type: *mut Schema_ComponentUpdate,
+// A data structure which represents all known component types. Used to generate an array of vtables to pass
+// to the connection object.
+pub struct ComponentDatabase {
+    component_vtables: Vec<worker::Worker_ComponentVtable>,
+}
+
+impl ComponentDatabase {
+    pub fn new() -> Self {
+        ComponentDatabase {
+            component_vtables: Vec::new(),
+        }
+    }
+
+    pub fn add_component<C: Component>(mut self) -> Self {
+        self.component_vtables.push(create_component_vtable::<C>());
+        self
+    }
+
+    pub(crate) fn to_worker_sdk(&self) -> *const worker::Worker_ComponentVtable {
+        self.component_vtables.as_ptr()
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.component_vtables.len()
+    }
+}
+
+// An object which contains a reference counted instance to T.
+struct ClientHandle<T> {
+    data: Rc<T>,
+    // When Drop() is called, data should reduce its reference count.
 }
 
-impl From<&Worker_ComponentUpdate> for ComponentUpdate {
-    fn from(update: &Worker_ComponentUpdate) -> Self {
-        ComponentUpdate {
-            component_id: update.component_id,
-            schema_type: update.schema_type,
+impl<T> ClientHandle<T> {
+    fn new(data: T) -> ClientHandle<T> {
+        ClientHandle {
+            data: Rc::new(data),
+        }
+    }
+
+    fn copy(&self) -> ClientHandle<T> {
+        ClientHandle {
+            data: Rc::clone(&self.data),
         }
     }
+
+    // Vtable helper functions.
+    unsafe fn handle_allocate(data: T) -> *mut raw::c_void {
+        // Allocate client handle and initialize with data.
+        let new_client_handle: *mut ClientHandle<T> =
+            mem::transmute(alloc::alloc(Layout::new::<ClientHandle<T>>()));
+        ptr::write(new_client_handle, ClientHandle::<T>::new(data));
+
+        // Return new pointer.
+        mem::transmute(new_client_handle)
```

Comment:
I think the logic for `handle_allocate` could be much simpler. I think all it needs to be is

```
Arc::into_raw(Arc::new(data)) as *mut _
```

Using [`into_raw`](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.into_raw) will give you the raw pointer, and will ensure that the counter isn't decremented until you use [`from_raw`](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.from_raw) to restore it.

Reply:
Also, it's worth noting that you don't have to use `transmute` for pointer conversions, you can cast between pointer types.

Reply:
Wow, that's nice. I wasn't aware of `into_raw` and `from_raw`. Does it essentially allocate space for refcount + T, and turn that memory block into a pointer to T (so the refcount is located at the memory address of T, minus a couple of bytes)?

Reply:
`Arc::new` allocates space for the refcount + `T`. `into_raw` simply calls `mem::forget` on the `Arc` (so that it doesn't get dropped and deallocate the data) and then returns a pointer to the data (you can see the implementation [here](https://doc.rust-lang.org/src/alloc/rc.rs.html#394-398)). When you call `from_raw`, it knows where the refcount is laid out relative to the data, so it can reconstruct the full `Arc` from just a pointer to the data (you can see the implementation [here](https://doc.rust-lang.org/src/alloc/rc.rs.html#429-444).